GpxExport/Models/Workout+FitWriter.swift

```diff
@@ -42,74 +42,80 @@ extension Workout {
                     currentHeartrateIndex += 1
-                    heartrateString = self.gpxHeartRate(currentHeartrate)
-                }
-                if let trackpoint = self.gpxTrackPoint(location: location, heartrate: heartrateString).data(using: .utf8) {
-                    file.write(trackpoint)
+                    heartrate = UInt8(currentHeartrate)
                 }
+                let latitude = ValidatedMeasurement(value: location.coordinate.latitude, valid: true, unit: UnitAngle.garminSemicircle)
+                let longitude = ValidatedMeasurement(value: location.coordinate.longitude, valid: true, unit: UnitAngle.garminSemicircle)
+                let altitude = ValidatedMeasurement(value: location.altitude, valid: true, unit: UnitLength.meters)
+                let record = RecordMessage(timeStamp: FitTime(date: location.timestamp),
+                                           position: Position(latitude: latitude, longitude: longitude),
+                                           distance: nil,
+                                           timeFromCourse: nil,
+                                           cycles: nil,
+                                           totalCycles: nil,
+                                           accumulatedPower: nil,
+                                           altitude: altitude,
+                                           speed: nil,
+                                           power: nil,
+                                           gpsAccuracy: nil,
+                                           verticalSpeed: nil,
+                                           calories: nil,
+                                           verticalOscillation: nil,
+                                           stanceTime: stanceTime,
+                                           heartRate: heartrate,
+                                           cadence: nil,
+                                           grade: nil,
+                                           resistance: nil,
+                                           cycleLength: nil,
+                                           temperature: nil,
+                                           activity: nil,
+                                           torqueEffectiveness: torqueEffectiveness,
+                                           pedalSmoothness: pedalSmoothness,
+                                           stroke: nil,
+                                           zone: nil,
+                                           ballSpeed: nil,
+                                           deviceIndex: nil)
+                messages.append(record)
             }
-            file.write("""
-  </trkseg>
-  </trk>
-  </gpx>
 
-  """.data(using: .utf8)!)
-            file.closeFile()
+            do {
+                let encoder = FitFileEncoder(dataValidityStrategy: .none)
+                let data = try encoder.encode(fildIdMessage: fileId, messages: messages)
```

Comment:
I would turn on the ValidityStrategy to FileType or GarminConnect.  Use the garminConnect option if you ever expect to send it up to GarminConnect in the Future, as they have some other Rules applied to Activity File type that are stricter then the standard Required fields for activity type files

Reply:
Along with that you should probably add in a DeviceInfoMessage.